jalib/jassert.cpp

```diff
@@ -46,6 +43,9 @@ int jassert_quiet = 0;
 static int theLogFileFd = -1;
 static int errConsoleFd = -1;
 
+static jalib::string& tmpDir() {static jalib::string s;return s;};
+static jalib::string& uniquePidStr() {static jalib::string s;return s;};
```

Comment:
I deleted some of the previous messages, which were distracting and due to my own misunderstanding of the technical issue involved here.
First, we agree that this usage of `foo() = "newValue";` with `foo()` containing a static local variable in cases where `foo()` would be declared static within the same file, and where `foo()` was just a function, and not a method inside some other class.
Second, a preferred option in an ideal world would be to promote the static local variable of `foo()` into a static global variable.  However, in C++, this would cause the global variable to be allocated, and any default constructor would be invoked, even before main() is called.  That creates problems if the constructor invokes any wrapper functions from DMTCP.  In C++, this initialization of the static local variable is deferred until the first call to `foo()`, and so it will now occur only after main() is called, and only after DMTCP has been initialized.

CONCLUSION:  Even though programmers new to C++ will initially be confused by seeing a function on the left-hand side, this is a necessary evil.  As long as we document the definition of `foo()` with a comment ahead of it, this will mostly avoid any confusion among programmers new to C++.